pkg/webhooks/mutator/virt-launcher-mutator.go

```diff
@@ -0,0 +1,234 @@
+	cli          client.Client
+	hcoNamespace string
+	decoder      *admission.Decoder
+	logger       logr.Logger
+}
+
+func NewVirtLauncherMutator(cli client.Client, hcoNamespace string) *VirtLauncherMutator {
+	return &VirtLauncherMutator{
+		cli:          cli,
+		hcoNamespace: hcoNamespace,
+		logger:       log.Log.WithName("virt-launcher mutator"),
+	}
+}
+
+func (m *VirtLauncherMutator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	m.logInfo("Starting virt-launcher mutator handling")
+
+	if req.Operation != admissionv1.Create {
+		m.logInfo("not a pod creation - ignoring")
+		return admission.Allowed(ignoreOperationMessage)
+	}
+
+	launcherPod := &k8sv1.Pod{}
+	err := m.decoder.Decode(req, launcherPod)
+	if err != nil {
+		m.logErr(err, "cannot decode virt-launcher pod")
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	originalPod := launcherPod.DeepCopy()
+
+	hco, err := getHcoObject(ctx, m.cli, m.hcoNamespace)
+	if err != nil {
+		m.logErr(err, "cannot get the HyperConverged object")
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	enforceCpuLimits, enforceMemoryLimits, err := m.getResourcesToEnforce(ctx, launcherPod.Namespace, hco)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	if !enforceCpuLimits && !enforceMemoryLimits {
+		return admission.Allowed(ignoreOperationMessage)
+	}
+
+	if err := m.handleVirtLauncherCreation(launcherPod, hco, enforceCpuLimits, enforceMemoryLimits); err != nil {
+		m.logErr(err, "failed handling launcher pod %s", launcherPod.Name)
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	allowResponse := m.getAllowedResponseWithPatches(launcherPod, originalPod)
+	m.logInfo("mutation completed successfully for pod %s", launcherPod.Name)
+	return allowResponse
+}
+
+func (m *VirtLauncherMutator) handleVirtLauncherCreation(launcherPod *k8sv1.Pod, hco *v1beta1.HyperConverged, enforceCpuLimits, enforceMemoryLimits bool) error {
+	var cpuRatioStr, memRatioStr string
+	if enforceCpuLimits {
+		cpuRatioStr = hco.Annotations[cpuLimitToRequestRatioAnnotation]
+	}
+	if enforceMemoryLimits {
+		memRatioStr = hco.Annotations[memoryLimitToRequestRatioAnnotation]
+	}
+
+	if enforceCpuLimits {
+		err := m.setResourceRatio(launcherPod, cpuRatioStr, cpuLimitToRequestRatioAnnotation, k8sv1.ResourceCPU)
+		if err != nil {
+			return err
+		}
+	}
+	if enforceMemoryLimits {
+		err := m.setResourceRatio(launcherPod, memRatioStr, memoryLimitToRequestRatioAnnotation, k8sv1.ResourceMemory)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (m *VirtLauncherMutator) setResourceRatio(launcherPod *k8sv1.Pod, ratioStr, annotationKey string, resourceName k8sv1.ResourceName) error {
+	ratio, err := strconv.ParseFloat(ratioStr, 64)
+	if err != nil {
+		return fmt.Errorf("%s can't parse ratio %s to float: %w. The ratio is the value of annotation key %s", launcherMutatorStr, ratioStr, err, annotationKey)
+	}
+
+	if ratio <= 0 {
```

Comment:
i would even say that a ratio < 1 is not supported, because a limit that is smaller than the request makes no sense.